### PR TITLE
FIR checker: report SETTER_PROJECTED_OUT

### PR DESCRIPTION
**Pull request by tgeng**
FE1.0 stores a boolean that signifies whether a (fake var) property has the
setter projected out at
https://github.com/JetBrains/kotlin/blob/d1fd1da56f222741c1626d4520128b2e30efc83f/core/descriptors/src/org/jetbrains/kotlin/descriptors/impl/PropertyDescriptorImpl.java#L466.

In this PR, instead of storing such a boolean in FirProperty, we simply
check if the LHS type is a captured type and this seems to be sufficient.

### CODE REVIEW

File: compiler/testData/diagnostics/tests/generics/varProjection/setterProjectedOutAssign.fir.kt [hunk 1]
```diff
@@ -7,7 +7,7 @@ interface Tr<T> {
 
 fun test(t: Tr<*>) {
     t.v = null!!
-    t.v = <!ASSIGNMENT_TYPE_MISMATCH!>""<!>
-    t.v = <!NULL_FOR_NONNULL_TYPE!>null<!>
+    t.v = <!SETTER_PROJECTED_OUT!>""<!>
```
Comment (punzki) [inline compiler/testData/diagnostics/tests/generics/varProjection/setterProjectedOutAssign.fir.kt:10]: This is reported on `t.v` in FE 1.0. Then they would be `FIR_IDENTICAL`. I believe you should be able to use `SourceElementPositioningStrategies.ASSIGNMENT_LHS`

Comment (tgeng) [reply compiler/testData/diagnostics/tests/generics/varProjection/setterProjectedOutAssign.fir.kt:10]: Aha! Thank you!

Comment (jsjeon) [reply compiler/testData/diagnostics/tests/generics/varProjection/setterProjectedOutAssign.fir.kt:10]: The other test below (`setterProjectedOutNoPlusAssign`) is `FIR_IDENTICAL`, but this one is still not. I saw you updated the positioning strategy above; is this test data not updated?

**Comment by jsjeon** [comment]
Besides two changes in this PR, I wonder this one can be also covered? https://punzki.github.io/Kotlin-DiagnosticTestsDiff/2021-05-25/diff-tests__generics__varProjection__setterNotProjectedOutAssign.kt.html

**Comment by tgeng** [comment]
> Besides two changes in this PR, I wonder this one can be also covered? https://punzki.github.io/Kotlin-DiagnosticTestsDiff/2021-05-25/diff-tests__generics__varProjection__setterNotProjectedOutAssign.kt.html

I think FE1.0's behavior is actually incorrect. In this case, the field is not projected out because the type `Tr<*>` is not a projection: it has a projection type parameter, but this type parameter does not "remove" setter for `v`. The name of the test file `setterNotProjectedOutAssign.kt` actually describes this situation and FE1.0's behavior is conflicting with this name.

**Review by jsjeon — APPROVED** [review]
Thanks for the explanation, that makes sense!

File: compiler/testData/diagnostics/tests/generics/varProjection/setterProjectedOutAssign.fir.kt (hunk 1 above)
Comment (tgeng) [reply compiler/testData/diagnostics/tests/generics/varProjection/setterProjectedOutAssign.fir.kt]: It turns out I need to report the source of the whole assignment instead. Fixed and updated.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/FirHelpers.kt [hunk 2]
```diff
@@ -523,6 +520,7 @@ private val FirSimpleFunction.matchesToStringSignature: Boolean
 
 fun checkTypeMismatch(
     lValueOriginalType: ConeKotlinType,
+    assignment: FirVariableAssignment?,
     rValue: FirExpression,
     context: CheckerContext,
     source: FirSourceElement,
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/FirHelpers.kt:526]: The `source` parameter here is actually `rValue.source` so you can just remove it so it's not confusing which element's source it is, and just use `rValue.source` everywhere (or a local variable for `rValue.source`).

**Review by punzki — APPROVED** [review]

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually 👍

**Pull request closed without merging**